Validate API input & Generate documentation (swagger) using tsoa

Nikita-Polyakov opened the pull request:
Issue: https://github.com/traittech/actiongate/issues/2

mezrin commented:
нужно добавить одну вещь при валидации структуры clearing transaction

---

При создании клиринговой транзакции указывается для какого AppAgent она выполняется
Поле appAgentId

```
export interface ClearingTransactionPayload {
  signatory: string;
  appAgentId: number;
  atomics: CTAtomic[];
}
```

Внутри экшенов указывается кто является origin экшена

```
export type CTActionOrigin =
  | {
      /** The ID of the application agent. */
      AppAgentId: BlockchainGenericId
    }
  | {
      /** The address of the application agent. */
      AppAgentAddress: BlockchainGenericAccount
    }
  | {
      /** The ID of the transactional address. */
      TransactionalAddressId: BlockchainGenericId
    }
  | {
      /** The transactional address. */
      TransactionalAddress: BlockchainGenericAccount
    }
  | {
      /** The named address. */
      NamedAddress: BlockchainGenericAccount
    }
  | {
      /** The name of the named address. */
      NamedAddressName: any
    };
```

Так вот origin экшена - это всегда keyless address, который контролирует AppAgent, указанный в ClearingTransactionPayload.
В противном случае транзакция закончится с ошибкой.

---

Нужно выполнять проверку

если `action_origin === CTActionOrigin.TransactionalAddressId OR CTActionOrigin.NamedAddressName`
ничего не делать

если `action_origin === CTActionOrigin.AppAgentId`
проверить что action_origin == ClearingTransactionPayload.appAgentId

если `action_origin === CTActionOrigin.AppAgentAddress OR CTActionOrigin.TransactionalAddress OR CTActionOrigin.NamedAddress`
- сделать декод адреса с помощью либы trait-keyless, доступно в NPM
- проверить, что appAgentID, закодированный в адрес, равен ClearingTransactionPayload.appAgentId

mezrin commented:
Метаданные от ноды содержат комментарии к экстринсикам
Ниже скриншот для transfer_allow_death

Надо добавить эти описания к генерируемым моделям
Из примера выше - для CTAtomicAction.Balances.TransferAllowDeath

<img width="704" alt="Screenshot 2025-01-21 at 16 25 55" src="https://github.com/user-attachments/assets/41bab23a-7fed-4a0d-b779-8c79e88e0018" />

mezrin commented:
The endpoint /submit/transaction can process the all extrinsics enabled on the blockchain

- pallets balances / nfts / assets - same as for the endpoint /submit/clearing_transaction
- other pallets - all extrinsics

Nikita-Polyakov commented:
> The endpoint /submit/transaction can process the all extrinsics enabled on the blockchain
> 
> * pallets balances / nfts / assets - same as for the endpoint /submit/clearing_transaction
> * other pallets - all extrinsics

I've started to adding methods from other pallets
https://github.com/traittech/actiongate/pull/7/commits/be52979091cb972c97bd43ab3b43b1d8ddfd2c2d

Did you mean that i should add all the methods from the other blockchain pallets?

mezrin reviewed (changes requested):
the first batch of requests for change

mezrin commented:
Some extrinsics use MultiAddress as a parameter - it allows to use address / int ID and other formats of addresses.
In TRAIT we are planning to use only regular Substrate ss58 addresses.
But we didn't delete MultiAddress from everywhere in blockchain node because it's just too much work and boilerplate.
As a result API calls look like this

```
{
  "signatory": "ADMIN_1",
  "appAgentId": 4294967295,
  "atomics": [
    {
      "actions": [
        {
          "actionType": "balances.transferAllowDeath",
          "arguments": {
            "dest": {
              "id": "ttxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
            },
            "value": "0416412833087280096978208"
          },
          "origin": {
            "AppAgentId": 4294967295
          }
        }
      ]
    }
  ]
}
```

put attention on argument `dest`

```
"dest": {
  "id": "ttxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
},
```

This format will confuse users of ActionGate who are not very familiar with Substrate.

We need to change this.

- API  params shall look like this: `"dest": "ttxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"`. We expect only regular ss58 address here
- when we encode the arguments of the call we convert it to the needed form: ""dest": { "id": "ttxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" },"

mezrin commented:
When we make a call to the service and api keys are enabled - the request must contain the header X-API-Key

`src/docs/swagger.json`
This file specifies Swagger UI and gets rebuilt on every "yarn build".
At the moment this file contains empty specs for security of endpoints: `"security": [],`.
Need to fix it

mezrin commented:
> > The endpoint /submit/transaction can process the all extrinsics enabled on the blockchain
> > 
> > * pallets balances / nfts / assets - same as for the endpoint /submit/clearing_transaction
> > * other pallets - all extrinsics
> 
> Did you mean that i should add **all** the methods from the other blockchain pallets?
> 
> I'm not sure that i should add something from `system`, `sudo`, `polkadotXcm`, `timestamp`, `parachainSystem` and other default system pallets (i guess they are not in API use cases).
> 
> I've added methods from TRAIT pallets to support them for `submit/transaction` endpoint:
> 
> * `admins`
> * `appAgents`
> * `appResources`
> * `appSubscriptions`
> * `appTransactions`
> * `appTransferChannels`
> * `appTransferFilters`
> * `userFreeTransactions`
> * `userTransactions`
> * `userVerification`
> 
> Could you clarify please, which pallets I should add too?

Checked the current sate of the API. 

## endpoint `/submit/clearing_transaction`

add to the list of available actions

`appTransferChannels.establishTransferChannel`
`appTransferChannels.removeTransferChannel`



## endpoint `/submit/transaction`

Only the following transactions shall be available
(the rest of currently available extrinsics are for blockchain root only.
Or for keyless addresses and, as a result, are callable only via a clearing transaction)

AppAgentsAddAdminTx
AppAgentsAddAdminToNamedAddressTx
AppAgentsAllowAdminColdWalletTx
AppAgentsChangeHotWalletTx
AppAgentsChangeOwnerCancelTx
AppAgentsChangeOwnerCompleteTx
AppAgentsChangeOwnerInitTx
AppAgentsClearAdminDispatchFilterTx
AppAgentsClearAppAgentMetadataTx
AppAgentsCompleteDestroyAppAgentTx
AppAgentsDisableAdminColdWalletDispatchTx
AppAgentsDisableHotWalletTx
AppAgentsEnableHotWalletTx
AppAgentsInitiateDestroyAppAgentTx
AppAgentsPauseAppAgentTx
AppAgentsProcessDestroyTx
AppAgentsReactivateAppAgentTx
AppAgentsRemoveAdminTx
AppAgentsRemoveAdminFromNamedAddressDispatchTx
AppAgentsSetAdminDispatchFilterTx
AppAgentsSetAppAgentMetadataTx
AppAgentsUnpauseAppAgentTx

AppSubscriptionsSetAppPayOnDemandModeTx
AppSubscriptionsSetAppSubscriptionTierTx

AppTransferFiltersBlockAddressDepositsTx
AppTransferFiltersSetAppDepositPermissionsTx
AppTransferFiltersUnblockAddressDepositsTx

AssetsTransferTx
AssetsTransferKeepAliveTx

BalancesTransferAllTx
BalancesTransferAllowDeathTx
BalancesTransferKeepAliveTx

NftsTransferItemTx

mezrin commented:
Let's please rename the action/transaction in API

BalancesTransferAllowDeathAction -> BalancesTransferAction
`AllowDeath` is confusing for an engineer who is not familiar with Substrate

Nikita-Polyakov commented:
> When we make a call to the service and api keys are enabled - the request must contain the header X-API-Key
> 
> `src/docs/swagger.json` This file specifies Swagger UI and gets rebuilt on every "yarn build". At the moment this file contains empty specs for security of endpoints: `"security": [],`. And on every rebuild with `yarn build` the file gets rebuilt and changes in endpoint security gets wiped. Need to fix it

Was fixes in [commit ](https://github.com/traittech/actiongate/pull/7/commits/cae161dbf5558e805132d11b0ef4cd373c2eb72f)
For this, i used `@Security()` decorator from `tsoa`, and added `expressAuthentication` as described in [tsoa docs](https://tsoa-community.github.io/docs/authentication.html). The logic from `checkApiKey` middleware was transfered here. After that i removed `checkApiKey` middleware, because now our routes has security decorators.

Nikita-Polyakov commented on src/validator/checks.ts:
```diff
@@ -0,0 +1,11 @@
+  } catch {
+    return false;
+  }
+};
```
I updated `isValidSS58 ` method, and also added a pattern to `BlockchainGenericAccount`

Lets take a look to the `ttxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx` address as argument:

1) Validation by pattern `^tt[1-9a-zA-Z]{47}$` executes first using `tsoa `tools. But this test address will pass this regexp.
2) Validation by `zod `library that calls  `isValidSS58` throws error.

This case is described in `clearing-tx-builder.test.ts`

![ss58](https://github.com/user-attachments/assets/c1c9f5e0-8d8d-4349-a0b1-2555912946fc)

Nikita-Polyakov commented on src/types/api/trait/appAgents.ts:
```diff
@@ -0,0 +1,49 @@
+ *
+ * [WARNING]: `openapi v3` doesn't support `tuple` type.
+ */
+type NonFungibleDestroyDetail = UINT32[];
```
Resolved this issue using argument transform on validation step in this [commit](https://github.com/traittech/actiongate/pull/7/commits/ae15d168e44fbfafa2cccf1b1985911f339c43c8)

User passes this argument as object, then after validation it becomes tuple

Nikita-Polyakov commented on src/types/api/clearingTransaction.ts:
```diff
@@ -1,138 +1,103 @@
+ *
+ * [WARNING]: `openapi v3` doesn't support `tuple` type.
+ */
+export type CTAction = Array<CTActionOrigin | CTActionCall>;
```
After `Txs`& `Actions` lists were updated, i returned `tuple `type, because it's not used in openapi documentation now

mezrin commented:
have some bug in building the transactions

Input
```
curl -X 'POST' \
  'http://localhost:8080/submit/transaction' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H 'X-API-KEY: api-key-2' \
  -d '{
  "signatory": "FAUCET",
  "tx": {
    "actionType": "balances.transferAllowDeath",
    "arguments": {
      "dest": "ttmojTij44xvCLsMZ1KHEyRfgcc26aJVdiy8xttuyoUQ8Li8s",
      "value": "1000000000000"
    }
  }
}'
```

Output
```
{"status":"failed","error_code":500,"error_description":"Internal Server Error","error_details":"Cannot read properties of undefined (reading 'AppAgentsAddAdmin')"}
```

In the logs
```
2025-03-17T21:13:27.767Z info: >> submitTransaction endpoint
2025-03-17T21:13:27.767Z info: Retrieving private key for the signatory...
2025-03-17T21:13:27.769Z info: Calling Account: ttpPFQ36WR4BCYto77ngSPThK4mN73RWzmof1oJjhDAjPux82
2025-03-17T21:13:27.769Z info: Creating and broadcasting signed transaction...
2025-03-17T21:13:27.769Z debug: Generating transaction metadata...
2025-03-17T21:13:27.770Z debug: Generating transaction metadata for address: ttpPFQ36WR4BCYto77ngSPThK4mN73RWzmof1oJjhDAjPux82
2025-03-17T21:13:27.770Z debug: Retrieving build information for address: ttpPFQ36WR4BCYto77ngSPThK4mN73RWzmof1oJjhDAjPux82...
2025-03-17T21:13:28.358Z debug: Build information retrieved successfully.
2025-03-17T21:13:28.433Z debug: Transaction metadata generated successfully
2025-03-17T21:13:28.433Z info: Building unsigned transaction for: balances.transferAllowDeath, with args: {"dest":"ttmojTij44xvCLsMZ1KHEyRfgcc26aJVdiy8xttuyoUQ8Li8s","value":"1000000000000"}
2025-03-17T21:13:28.433Z error: Error creating and broadcasting signed transaction: Cannot read properties of undefined (reading 'AppAgentsAddAdmin')
2025-03-17T21:13:28.433Z error: Error creating and broadcasting transaction: Cannot read properties of undefined (reading 'AppAgentsAddAdmin')
```

mezrin commented:
For me the SS58 check of the address still doesn't work

Request
```
curl -X 'POST' \
  'http://localhost:8080/submit/transaction' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H 'X-API-KEY: api-key-2' \
  -d '{
  "signatory": "FAUCET",
  "tx": {
    "actionType": "balances.transferAllowDeath",
    "arguments": {
      "dest": "ttxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      "value": "1000000000000"
    }
  }
}'
```

Output
```
{"status":"failed","error_code":500,"error_description":"Internal Server Error","error_details":"Cannot read properties of undefined (reading 'AppAgentsAddAdmin')"}
```

Basically this error means that app starts to build the TX.
But it should return a error much earlier - on parsing&validation the input.

mezrin commented:
We have some issue with parsing of validation of uint128

Input
```
curl -X 'POST' \
  'http://localhost:8080/submit/transaction' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H 'X-API-KEY: api-key-2' \
  -d '{
  "signatory": "FAUCET",
  "tx": {
    "actionType": "balances.transferAllowDeath",
    "arguments": {
      "dest": "ttxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      "value": "999999999999999999999999999999999999999"
    }
  }
}'
```

You see the value:
`999999999999999999999999999999999999999`
But the max uint128 is
`340282366920938463463374607431768211455`


Output
```
{"status":"failed","error_code":500,"error_description":"Internal Server Error","error_details":"Cannot read properties of undefined (reading 'AppAgentsAddAdmin')"}
```

This means that `999999999999999999999999999999999999999` passed the validation of the input
Need to figure out it

mezrin reviewed (changes requested):
The code looks good! For the merge, need to fix the issues mentioned above.

Nikita-Polyakov commented:
I don't have this weird error in my build, it looks like in every case above this error causes of status `500`.

`{"status":"failed","error_code":500,"error_description":"Internal Server Error","error_details":"Cannot read properties of undefined (reading 'AppAgentsAddAdmin')"}`

I honestly don't know what the reason could be, because judging by the error message, the `ExtrinsicType `variable is `undefined`. This looks like a broken build or outdated code.

Could you please to check these steps to prepare build?
1) Run `yarn` command to install packages
2) Create `config.json` file - we can test the application behavior with this configuration.

```
{
  "api_keys_enabled": true,
  "api_keys": ["api-key-1"],
  "api_ssl_enabled": false,
  "api_ssl": {
    "sslKeyPath": "/path/to/ssl/key.pem",
    "sslCertPath": "/path/to/ssl/cert.pem"
  },
  "datagate_api": {
    "uri": "https://api.trait.tech"
  },
  "tx_build": {
    "tx_life_time": 64
  },
  "blockchain": {
    "ss58_code": 5335
  },
  "admin_keys": [
    {
      "id": "ADMIN",
      "private_key": "street firm worth record skin taste legend lobster magnet stove drive side"
    }
  ],
  "logs": {
    "logLevel": "debug",
    "logFile": "actiongate.log"
  }
}
```

3) Run `yarn build`
4) Run `yarn start`

I haven't tokens in this test account `ttqRZiq3D4jbPMCDzsWV5PmdXCH3rEtChCkfgfA6CXYQQyYKb` right now, but transactions are sending to datagate and i see the error that my balance is too low to pay fees.

```
Submit response: {"metadata":{"request_context":{"block_hash":"0xcbddb57336d1
3d4d01ac86fab8707efa8357198cca24ad253c1cbed7357685b7","block_index":86179,"block_timestamp":1742303136000},"b
lockchain_head":{"block_hash":"0xcbddb57336d13d4d01ac86fab8707efa8357198cca24ad253c1cbed7357685b7","block_ind
ex":86179,"block_timestamp":1742303136000}},"data":{"acceptance":false,"success_info":null,"rejection_info":{
"rejection_code":"BLOCKCHAIN_AUTHORSHIP__POOL_INVALID_TX","rejection_text":"Transaction pool rejected the tra
nsaction as invalid","blockchain_error":{"code":1010,"message":"Invalid Transaction","data":"Inability to pay
 some fees (e.g. account balance too low)"}}}}
```

mezrin commented:
I launched the app with `yarn run build && yarn run start` I had that error.
Now I deleted the `dist` dir and launched again - all works.
Indeed some build issues

mezrin commented:
all works great! thank you! merging the PR
